# pan-sol-spec (beta)
The pan-sol specification provides details for naming, formatting, and versioning plant pan-genomic resources. Broadly, the specification provides details for organizing diverse plant samples within a species and their associated genome assembly and annotation files. The specification is especially well-suited for new species with few legacy files/conventions.

**accounting for structural diversity**

This specification envisions a collection of "reference" or "platinum" quality eukaryotic genome assemblies. For convenience, one (the first for species without an existing reference) assembly is considered _the_ reference genome. The collection of all assemblies establishes an abstract pan-genome, easily amenable to graph-based and other multi-reference pan-genome analyses. Aside from multi-reference analyses, any individual assembly could be used as a "reference", enabling "personalized" genomics analyses.

**accounting for genes**

It is inconvenient to treat annotations independently for each sample within a species. Instead, this specification maintains a database of all known genes in the species. This database is initialized with all of the genes in *the* reference genome and is updated to include new genes as they are discovered.  When annotating genes in a new sample, known genes are annotated with their previously assigned ID/Names. If a gene is found in >1 copy, the ID is updated accordingly. This accounts for presence/absence variation while maintaining consistent information about known genes. 




## Samples

Below provides information for a species and distinct accessions/haplotypes within a species.

### Definitions

#### Species Code

The first three letters of the genus and the first three letters of the species. For example, for *Solanum lycopersicum*, the species code would be `Sollyc`. If a code is already taken within the genus, any unique 6 character code can be made instead. For example, for *Solanum lycopersicoides*, one might choose the code `Solcoi`.

#### Accession Code

A unique integer assigned to a particular accession.

#### Accession Haplotype

When phasing is necessary, the terms `mat` and `pat` can be used to distinguish resources for the maternal and paternal haplotypes, respectively.

## Assemblies

Below provides inormation for individual genome assemblies

### Definitions

#### Assembly Name

For non-phased assemblies, this is:

`{Species Code}` + `{Accession Code}` (e.g. `Sollyc1`)

For phased assemblies, add the Accession Haplotype (e.g. `Sollyc1mat`).

#### Sequence

An individual record in an assembly fasta file

#### Sequence Name

A unique sequence name, used as the FASTA header

#### Sequence Group

A collection of sequences

### Sequence Name/Group Assignment




| sequence                                      | sequence name	                         | sequence name regex | sequence group    | example         |
| :---                                          |    :----                               |         :---        |  :---             | :---            |
| nuclear chromosomes                           | `chrN` \*                              |  `^chr\d{2}$` \*\*  |  CHROMOSOME_N     |  `chr1`         |
| unplaced nuclear scaffolds                    | `scf` + unique 8 digit integer         |  `^scf\d{8}$`       |  UNPLACED_NUCLEAR |  `scf00000001`  |
| unplaced nuclear contigs                      | `ctg` + unique 8 digit integer         |  `^ctg\d{8}$`       |  UNPLACED_NUCLEAR |  `ctg00000001`  |
| chloroplast genome                            | `chrC`                                 |  `^chrC$`           |  CHLOROPLAST      |  `chrC`         |
| mitochondrial genome molecule/scaffold/contig | `chrM` + unique 8 digit integer \*\*\* |  `^chrM_\d$`        |  MITOCHONDRION    |  `chrM00000001` |

> \* where `N` is the chromosome number
> \*\* assuming two digits are needed to represent the largest chromosome number
> \*\*\* if only a single master molcule is included, one can name it 'chrM' (`^chrM$`)

### FASTA formatting and versioning

#### Versioning
Assembly version numbers are formated as BUILD.PATCH, and the first production-ready FASTA file is v1.0. When updating the assembly, increment the:

1. BUILD version when the update induces a coordinate system change of existing sequences,
2. PATCH version when the update does not induce a change to the coordinate system of existing sequences.

#### FASTA File Name
`{Assembly Name}` + `_` + `vBUILD.PATCH` + `.fasta`

For example, `Sollyc1_v1.0.fasta`



## Annotations