# pan-sol-spec (beta)
The pan-sol specification provides details for naming, formatting, and versioning plant pan-genomic resources. Broadly, the specification provides details for organizing diverse plant samples within a species and their associated genome assembly and annotation files. The specification is especially well-suited for new species with few legacy files/conventions.

**accounting for structural diversity**

This specification envisions a collection of "reference" or "platinum" quality eukaryotic genome assemblies. For convenience, one (the first for species without an existing reference) assembly is considered _the_ reference genome. The collection of all assemblies establishes an abstract pan-genome, easily amenable to graph-based and other multi-reference pan-genome analyses. Aside from multi-reference analyses, any individual assembly could be used as a "reference", enabling "personalized" genomics analyses.

**accounting for genes**

It is inconvenient to treat gene annotations independently for each sample within a species. Instead, this specification maintains a database of all known genes in the species. This database is initialized with all of the genes in *the* reference genome and is updated to include new genes as they are discovered.  When annotating genes in a new sample, known genes are annotated with their previously assigned ID/Names. If a gene is found in >1 copy, the ID is updated accordingly. This accounts for presence/absence variation while maintaining consistent information about known genes.

## Samples

Below provides information for a species and distinct accessions/haplotypes within a species.

### Definitions

| term                | definition	                                                                                                               |
| :---                | :---                                                                                                                       |
| Species Code \*     | The first three letters of the genus and the first three letters of the species. E.g. `Sollyc` for *Solanum lycopersicum*. | 
| Accession Name      | A name assigned to a particular accession.                                                                                 | 
| Accession Code      | A unique integer assigned to a particular accession.                                                                       | 
| Accession Haplotype | A code to distinguish resources for the maternal(`mat`) and paternal (`pat`) haplotypes.                                   | 

> \* If a code is already taken within the genus, any unique 6 character code can be made instead. For example, for *Solanum lycopersicoides*, one might choose the code `Solcoi`.

## Assemblies

Below provides information for individual genome assemblies

### Definitions

| term           | definition	                                                                                                                   |
| :---           | :---                                                                                                                            |
| Assembly Name  | `{Species Code}` + `{Accession Code}` (e.g. `Sollyc1`). For phased assemblies, add the Accession Haplotype (e.g. `Sollyc1mat`). | 
| Sequence       | An individual record in an assembly FASTA file.                                                                                 | 
| Sequence Name  | A unique sequence name, used as the FASTA identifier.                                                                           | 
| Sequence Group | A collection of sequences.                                                                                                      |

### Sequence Name/Group Assignment




| sequence                                      | sequence name	                         | sequence name regex | sequence group    | example         |
| :---                                          |    :----                               |         :---        |  :---             | :---            |
| nuclear chromosomes                           | `chrN` \*                              |  `^chr\d{2}$` \*\*  |  CHROMOSOME_N     |  `chr1`         |
| unplaced nuclear scaffolds                    | `scf` + unique 8 digit integer         |  `^scf\d{8}$`       |  UNPLACED_NUCLEAR |  `scf00000001`  |
| unplaced nuclear contigs                      | `ctg` + unique 8 digit integer         |  `^ctg\d{8}$`       |  UNPLACED_NUCLEAR |  `ctg00000001`  |
| chloroplast genome                            | `chrC`                                 |  `^chrC$`           |  CHLOROPLAST      |  `chrC`         |
| mitochondrial genome molecule/scaffold/contig | `chrM` + unique 8 digit integer \*\*\* |  `^chrM_\d$`        |  MITOCHONDRION    |  `chrM00000001` |

> \* where `N` is the chromosome number
>
> \*\* assuming two digits are needed to represent the largest chromosome number
> 
> \*\*\* if only a single master molcule is included, one can name it 'chrM' (`^chrM$`)

### FASTA formatting and versioning

#### Versioning
Assembly version numbers are formated as BUILD.PATCH, and the first production-ready FASTA file is `v1.0`. When updating the assembly, increment the:

1. BUILD version when the update induces a coordinate system change of existing sequences,
2. PATCH version when the update does not induce a change to the coordinate system of existing sequences.

#### FASTA File Name
`{Assembly Name}` + `_` + `vBUILD.PATCH` + `.fasta`

For example, `Sollyc1_v1.0.fasta`

#### FASTA File Sequence Headers

Headers start with a `>` character and are followed by the following space delimited fields:

1. Sequence Name
2. Assembly Name
3. Genus
4. Species
5. Accession Name
6. Accession Code
7. Sequence Group

For example:

`>chr1 Sollyc1 Solanum lycopersicum M82 1 CHROMOSOME_1`

#### FASTA File Sequences

Sequences are all upper case, with 60 character line-wrapping. Sequences should be sorted in the following order:

1. Nuclear chromosomes, numerically sorted
2. MITOCHONDRION group, numerically sorted (if available)
3. chrC (if available)
4. Unplaced contigs, numerically sorted
5. Unplaced scaffolds, numerically sorted

## Annotations

Below provides information for individual genome assembly annotations. This specification currently only addresses protein coding gene annotations.

### Gene Database
Genes are well-conserved within a typical plant species. Rather than treating gene annotations independently across intra-species accessions, it is prudent to maintain a species-wide annotation/database. In this scenario, assembly-specific annotations simply provide the gene coordinates and child features (mRNA, exon, CDS, and UTR) for a particular assembly. Still, this system must account for genic structural variation that is common in plant genomes.

### GFF3 formatting and versioning
Individual assembly annotations should be in the [GFF3 format](https://m.ensembl.org/info/website/upload/gff3.html).

#### Versioning
Annotation version numbers are formated as BUILD.MAJOR.MINOR, and the first production-ready GFF3 file is `v1.0.0`. The BUILD must match the corresponding assembly build. When updating the annotation, increment the:

1. BUILD version when the update provides an annotation to a new assembly BUILD (coordinate system),
2. MAJOR version for major annotation changes, such as the addition/removal/adjustment of many new features,
3. MINOR version for minor changes such as changes to a small number of features or to metadata/names

#### GFF3 File Name
`{Assembly Name}` + `_genes_` + `vBUILD.MAJOR.MINOR` + `.gff3`

For example, `Sollyc1_genes_v1.0.0.gff3`

#### Additional Files and Their Names
Proteins, CDS, and cDNA sequences should be provided in FASTA format for each GFF3 version. They should be named as follows:

| resource             | file name	                                                              | example                               | 
| :---                 | :---                                                                     | :---                                  |
| protein sequences \* | `{Assembly Name}` + `_genes_` + `vBUILD.MAJOR.MINOR` + `.proteins.fasta` | `Sollyc1_genes_v1.0.0.proteins.fasta` |
| CDS sequences        | `{Assembly Name}` + `_genes_` + `vBUILD.MAJOR.MINOR` + `.CDS.fasta`      | `Sollyc1_genes_v1.0.0.CDS.fasta`      |
| protein sequences    | `{Assembly Name}` + `_genes_` + `vBUILD.MAJOR.MINOR` + `.cDNA.fasta`     | `Sollyc1_genes_v1.0.0.cDNA.fasta`     |

> \* Terminal stop codons (often represented as `.`) should not be excluded from the FASTA file. Sequences with internal stop codons should have "INTERNAL STOP" in the associated FASTA header so that they could be easily removed if necessary.

### Gene IDs and Names

#### Accession ID
After establishing a draft annotation, each gene should be assigned a new attribute called an accession ID (`acc_id`). The accession ID is made up of 3 components:

##### The assembly name
The assembly name: e.g. `Sollyc1`

##### A chromosome code

The minimum number of digits needed to representt the largest chromosome number should be used for the chromosome code. It should also be followed by `g`. E.g. `02g` for the second chromosome of a genome with >9 and <99 chromosomes.

##### A 6 digit unique numeric ID
This specification does not require ordinal ID numbers. However, for ordinal numbers, the first gene number should be `000010` and subsequent numbers should be increments of 10.

Using these components, `Sollyc1.02g000010` is an example of the first gene on chromosome 2 of a tomato assembly.

#### Gene IDs

All default gene IDs must be replaced according to the following rules"

- If annotating the first accession in a species, the gene ID (`ID`) should be set to match the accession ID (`acc_ID`). The species gene database (referred to in the abstract) should then be initialized with these gene IDs and the accession that they were found in.

- If this is not the first accession to be annotated in the species, a homology-based or "lift over" annotation tool should be used to assign gene IDs for genes already present in the database.
  
- For any genes that are not already in the genes (novel genes), the gene ID is set to match the accession ID. These genes should then be added to the database, indicating which accession they were found in.

- If a known gene is found more than once, `-cN` should be added to the end of the gene ID, where `N` is the copy number of the gene.

#### Gene Names
Gene names can be added to the database at any time, and they should be assigned to the `Name` GFF3 attribute where possible.


